worker/dbaccessor/tracker.go

```diff
@@ -0,0 +1,244 @@
+	w.mutex.RLock()
+	defer w.mutex.RUnlock()
+
+	return w.err
+}
+
+// Kill implements worker.Worker
+func (w *trackedDBWorker) Kill() {
+	w.tomb.Kill(nil)
+}
+
+// Wait implements worker.Worker
+func (w *trackedDBWorker) Wait() error {
+	return w.tomb.Wait()
+}
+
+func (w *trackedDBWorker) loop() error {
+	timer := w.clock.NewTimer(PollInterval)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case <-timer.Chan():
+			// Any retryable errors are handled at the txn level. If we get an
+			// error returning here, we've either exhausted the number of
+			// retries or the error was fatal.
+			w.mutex.RLock()
+			currentDB := w.db
+			w.mutex.RUnlock()
+
+			newDB, err := w.verifyDB(currentDB)
+			if err != nil {
+				// If we get an error, ensure we close the underlying db and
+				// mark the tracked db in an error state.
+				w.mutex.Lock()
+				if err := w.db.Close(); err != nil {
+					w.logger.Errorf("error closing database: %v", err)
+				}
+				w.err = errors.Trace(err)
+				w.mutex.Unlock()
+
+				// As we failed attempting to verify the db, we're in a fatal
+				// state. Collapse the worker and if required, cause the other
+				// workers to follow suite.
+				return errors.Trace(err)
+			}
+
+			// We've got a new DB. Close the old one and replace it with the
+			// new one, if they're not the same.
+			if newDB != currentDB {
+				w.mutex.Lock()
+				if err := w.db.Close(); err != nil {
+					w.logger.Errorf("error closing database: %v", err)
+				}
+				w.db = newDB
+				w.err = nil
+				w.mutex.Unlock()
+			}
+		}
+	}
+}
+
+func (w *trackedDBWorker) verifyDB(db *sql.DB) (*sql.DB, error) {
+	// Force the timeout to be lower that the DefaultTimeout,
+	// so we can spot issues sooner.
+	// Also allow killing the tomb to cancel the context,
+	// so shutdown/restart can not be blocked by this call.
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Second*10)
+	ctx = w.tomb.Context(ctx)
+	defer cancel()
+
+	if w.logger.IsTraceEnabled() {
+		w.logger.Tracef("verifying database by attempting a ping")
+	}
+
+	// There are multiple levels of retries here. For good reason. We want to
+	// retry the transaction semantics for retryable errors. Then if that fails
+	// we want to retry if the database is at fault. In that case we want
+	// to open up a new database and try the transaction again.
+	for i := 0; i < DefaultVerifyAttempts; i++ {
+		// Verify that we don't have a potential nil database from the retry
+		// semantics.
+		if db == nil {
+			return nil, errors.NotFoundf("database")
+		}
+
+		err := database.Retry(ctx, func() error {
+			if w.logger.IsTraceEnabled() {
+				w.logger.Tracef("attempting ping")
+			}
+			return w.verifyDBFunc(ctx, db)
+		})
+		// We were successful at requesting the schema, so we can bail out
+		// early.
+		if err == nil {
+			return db, nil
+		}
+
+		// We failed to apply the transaction, so just return the error and
+		// cause the worker to crash.
```

Comment:
Are we actually applying a transaction here? AFAICT the only thing we are doing is calling verifyDBFunc which is likely just calling PingContext.
I *think* the strings here are incorrect.
Also, I think this comment is only true if the if returns true, vs the state that we are at now.

Reply:
We did originally, have better verification steps, but we move back to Ping. I'll update the documentation around this.